### PR TITLE
Migrate kubeflow-katib-presubmit to GitHub Actions

### LINKED ISSUE #1878: Migrate kubeflow-katib-presubmit to GitHub Actions
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
As described in https://github.com/kubeflow/katib/pull/1877#issuecomment-1140715656, we need to migrate our `kubeflow-katib-presubmit` test to Github Actions from kubeflow test-infra since it has been removed from the prow.

/cc @kubeflow/wg-automl-leads 

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

---

Love this feature? Give it a 👍 We prioritize the features with the most 👍

### PR DESCRIPTION
**Pull request by tenzen-y**
**What this PR does / why we need it**:
I migrated E2E tests from [kubeflow-testinfra](https://github.com/kubeflow/testing) to GitHub Actions with KinD Kubernetes Cluster since we no longer have access to  `kubeflow-katib-presubmit`.
Ref: https://github.com/kubeflow/katib/issues/1838

So, I made the following changes:

1. Moved old E2E test codes to `test/e2e/v1beta1/[scripts|hack]/aws` directory.
2. Created new E2E test codes in `test/e2e/v1beta1/[scripts|hack]/gh-actions` direcotory.
3. Built new workflows similar to `kubeflow-katib-presubmit` in GitHub Actions.
4. As discussed in our community meeting, reduced the following container images for Suggestion and EarlyStopping Services by changing the base image from python:3.9 to python:3.9-slim:
    - `cmd/earlystopping/medianstop/v1beta1/Dockerfile`
    - `cmd/metricscollector/v1beta1/tfevent-metricscollector/Dockerfile`
    - `cmd/suggestion/chocolate/v1beta1/Dockerfile`
    - `cmd/suggestion/hyperband/v1beta1/Dockerfile`
    - `cmd/suggestion/hyperopt/v1beta1/Dockerfile`
    - `cmd/suggestion/nas/darts/v1beta1/Dockerfile`
    - `cmd/suggestion/nas/enas/v1beta1/Dockerfile`
    - `cmd/suggestion/optuna/v1beta1/Dockerfile`
    - `cmd/suggestion/skopt/v1beta1/Dockerfile`
    - `examples/v1beta1/trial-images/enas-cnn-cifar10/Dockerfile.cpu`
5. Reduced the following container images for Trial containers by splitting container images by device types (CPU or GPU):
    - `examples/v1beta1/trial-images/darts-cnn-cifar10/Dockerfile.cpu`
    - `examples/v1beta1/trial-images/darts-cnn-cifar10/Dockerfile.gpu`
    - `examples/v1beta1/trial-images/pytorch-mnist/Dockerfile`
6. Changed parameters of examples to reduce required computer resources (ex., system memory, storage size, etc.).

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #1878 

/cc @andreyvelich @johnugeorge @gaocegege 

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

**Review by gaocegege — APPROVED** [review]
Thanks for your contribution! :tada: :+1:

File: examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt
Comment (johnugeorge) [inline examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt:2]: Ref: https://github.com/kubeflow/katib/pull/1870

File: examples/v1beta1/trial-images/pytorch-mnist/Dockerfile
Comment (johnugeorge) [inline examples/v1beta1/trial-images/pytorch-mnist/Dockerfile:2]: Why is the version different from the others?

File: examples/v1beta1/hp-tuning/tpe.yaml
Comment (tenzen-y) [reply examples/v1beta1/hp-tuning/tpe.yaml:52]: The Kubernetes Job retries 6 times by default, which increases the waiting time, so the `backoffLimit` is set to 0 to reduce the execution time.
Should I remove the setting for the `backoffLimit`?

File: .github/workflows/publish-algorithm-images.yaml
Comment (johnugeorge) [inline .github/workflows/publish-algorithm-images.yaml:29]: Is this added to see all errors at once?

File: examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt
Comment (tenzen-y) [reply examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt:2]: If we continue to wait for tensorflow-aarch64 v2.9.1, we can not use tensorflow v2.9.1 for Katib v0.14.0. 
So I will update `https://github.com/kubeflow/katib/pull/1870` to install protobuf (>= 3.9.2, < 3.20) to only container image for aarch64 after this PR merging.

File: examples/v1beta1/trial-images/pytorch-mnist/Dockerfile
Comment (tenzen-y) [reply examples/v1beta1/trial-images/pytorch-mnist/Dockerfile:2]: We need to fix the pytorch-mnist example to use Python v3.9 since both torch and torchvision can not execute in Python 3.9 as described. in [this](https://pypi.org/project/torchvision/).

> torch | torchvision | python
> -- | -- | --
> <=1.0.1 | 0.2.2 | ==2.7, >=3.5, <=3.7



So I will create another PR to upgrade the Python version after this PR merging.

File: .github/workflows/publish-algorithm-images.yaml
Comment (tenzen-y) [reply .github/workflows/publish-algorithm-images.yaml:29]: Yes.

File: examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt
Comment (johnugeorge) [reply examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt:2]: Should we merge #1870 first ? What do you think?

File: examples/v1beta1/hp-tuning/tpe.yaml
Comment (johnugeorge) [reply examples/v1beta1/hp-tuning/tpe.yaml:52]: Yes. Isn't this only when job is failed? Are we seeing job failure? Or do you mean, to fail the test immediately when job fails?

File: examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt
Comment (tenzen-y) [reply examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt:2]: SGTM.
I will send you a review request again after I update https://github.com/kubeflow/katib/pull/1870.

File: test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh
Comment (tenzen-y) [reply test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh:31]: These errors occurred in [this section](https://github.com/kubeflow/katib/blob/a9d92bd4a260816e3717d1a24f128ac65b461ace/Makefile#L2-L3).
So I think that we can ignore these errors.
WDYT? @johnugeorge

File: test/e2e/v1beta1/scripts/gh-actions/build-load.sh
Comment (johnugeorge) [inline test/e2e/v1beta1/scripts/gh-actions/build-load.sh:75]: If i understand correctly, core images will be built for every job while  suggestion/early stopping images will be built if used in the experiment.
Right? Is there any other way that we can avoid duplicate build? 

Is this https://github.com/docker/build-push-action/blob/master/docs/advanced/share-image-jobs.md helpful in this case?

File: test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh
Comment (johnugeorge) [reply test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh:31]: Ok. We can move this to make command later

File: test/e2e/v1beta1/scripts/gh-actions/build-load.sh
Comment (johnugeorge) [reply test/e2e/v1beta1/scripts/gh-actions/build-load.sh:75]: Let me know your thoughts on it. /cc @andreyvelich

File: test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh
Comment (tenzen-y) [reply test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh:31]: I will create another PR to fix Makefile.

File: test/e2e/v1beta1/scripts/gh-actions/build-load.sh
Comment (johnugeorge) [reply test/e2e/v1beta1/scripts/gh-actions/build-load.sh:75]: we can relook at optimising it later.

**Comment by johnugeorge** [comment]
Thanks @tenzen-y for this hard work

File: test/e2e/v1beta1/scripts/gh-actions/build-load.sh
Comment (tenzen-y) [reply test/e2e/v1beta1/scripts/gh-actions/build-load.sh:75]: > If i understand correctly, core images will be built for every job while suggestion/early stopping images will be built if used in the experiment.

Yes, that's right. @johnugeorge 

> Is this https://github.com/docker/build-push-action/blob/master/docs/advanced/share-image-jobs.md helpful in this case?

SGTM
I did not know how to do that.
Thanks! @johnugeorge

**Comment by tenzen-y** [comment]
> Thanks @tenzen-y for this hard work

Thank you for your comments and review.

File: test/e2e/v1beta1/scripts/gh-actions/build-load.sh
Comment (tenzen-y) [reply test/e2e/v1beta1/scripts/gh-actions/build-load.sh:75]: > we can relook at optimising it later. This looks good.

After merging this PR, I would create another PR to optimize the build script.
WDYT? @johnugeorge

Comment (johnugeorge) [reply test/e2e/v1beta1/scripts/gh-actions/build-load.sh:75]: Sure Thanks @tenzen-y

File: examples/v1beta1/hp-tuning/tpe.yaml
Comment (tenzen-y) [reply examples/v1beta1/hp-tuning/tpe.yaml:52]: Done.

File: examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt
Comment (tenzen-y) [reply examples/v1beta1/trial-images/enas-cnn-cifar10/requirements.txt]: Done.

File: examples/v1beta1/trial-images/pytorch-mnist/Dockerfile
Comment (tenzen-y) [reply examples/v1beta1/trial-images/pytorch-mnist/Dockerfile:2]: FYI: https://github.com/kubeflow/katib/issues/1883

**Comment by tenzen-y** [comment]
@johnugeorge  Please take another look.

/hold cancel

File: test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh
Comment (tenzen-y) [reply test/e2e/v1beta1/scripts/gh-actions/setup-katib.sh:31]: I have created PR to fix this.
https://github.com/kubeflow/katib/pull/1884

**Comment by johnugeorge** [comment]
Sure. I will keep it open for 1-2 more days so that others can take a look at it.

**Comment by tenzen-y** [comment]
> Sure. I will keep it open for 1-2 more days so that others can take a look at it.

It makes sense.

**Comment by tenzen-y** [comment]
I have pushed to rerun Go Test.

**Review by gaocegege — COMMENTED** [review]
LGTM!

**Comment by johnugeorge** [comment]
Thanks @tenzen-y 
/lgtm

**Pull request merged**
